packages/preferences/src/browser/preferences-widget.ts

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright (C) 2018 Red Hat, Inc. and others.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+import { inject } from "inversify";
+import { Message } from "@phosphor/messaging";
+import { PreferencesMenuFactory } from "./preference-menu-factory";
+import { PreferencesDecorator } from "./preferences-decorator";
+import {
+    ApplicationShell,
+    ContextMenuRenderer,
+    ExpandableTreeNode,
+    PreferenceSchemaProvider,
+    SelectableTreeNode,
+    PreferenceProperty,
+    PreferenceScope,
+    PreferenceService,
+    TreeModel,
+    TreeNode,
+    TreeProps,
+    TreeWidget
+} from "@theia/core/lib/browser";
+
+export abstract class PreferencesTreeWidget extends TreeWidget {
+
+    protected abstract handlePropertySelection: (property: string, value: any) => void;
+    private preferencesGroupNames: string[] = [];
+    private readonly properties: {[name: string]: PreferenceProperty};
+
+    @inject(PreferencesMenuFactory) protected readonly preferencesMenuFactory: PreferencesMenuFactory;
+    @inject(ApplicationShell) protected readonly applicationShell: ApplicationShell;
+    @inject(PreferenceService) protected readonly preferenceService: PreferenceService;
+
+    protected constructor(@inject(TreeModel) readonly model: TreeModel,
+                          @inject(TreeProps) protected readonly treeProps: TreeProps,
+                          @inject(PreferencesDecorator) protected readonly decorator: PreferencesDecorator,
+                          @inject(ContextMenuRenderer) protected readonly contextMenuRenderer: ContextMenuRenderer,
+                          @inject(PreferenceSchemaProvider) protected readonly preferenceSchemaProvider: PreferenceSchemaProvider) {
+        super(treeProps, model, contextMenuRenderer);
+        this.addClass('theia-preferences');
+
+        this.id = "theia-preferences-container";
+        this.title.label = this instanceof UserPreferencesTreeWidget ? 'User Preferences' : 'Workspace Preferences';
+        this.title.closable = true;
+        this.title.iconClass = 'fa fa-sliders';
+
+        this.properties = this.preferenceSchemaProvider.getCombinedSchema().properties;
+        for (const property in this.properties) {
+            if (property) {
+                const group: string = property.substring(0, property.indexOf('.'));
+                if (this.preferencesGroupNames.indexOf(group) < 0) {
+                    this.preferencesGroupNames.push(group);
+                }
+            }
+        }
+    }
+
+    protected onAfterAttach(msg: Message): void {
+        this.initializeModel();
+        super.onAfterAttach(msg);
+    }
+    protected onCloseRequest(msg: Message): void {
+        this.applicationShell
```

Comment:
We should compose the two widgets (tree & editor) in one instead of opening the editor on the right panel.

Reply:
done